googledataprocauthenticator/google.py

```diff
@@ -0,0 +1,319 @@
+from google.auth.exceptions import DefaultCredentialsError, RefreshError
+from hdijupyterutils.ipywidgetfactory import IpyWidgetFactory
+from hdijupyterutils.ipythondisplay import IpythonDisplay
+from google.oauth2.credentials import Credentials
+
+
+# The name of the Cloud SDK shell script
+_CLOUD_SDK_POSIX_COMMAND = "gcloud"
+_CLOUD_SDK_WINDOWS_COMMAND = "gcloud.cmd"
+# The command to get all credentialed accounts 
+_CLOUD_SDK_USER_CREDENTIALED_ACCOUNTS_COMMAND = ("auth", "list", "--format", "json")
+
+def load_json_input(result):
+    """Load json data from the file."""
+    jsondata = None
+    try:
+        jsondata = json.loads(result)
+    except:
+        raise
+    return jsondata
+
+def list_accounts_pairs(credentialed_accounts, default_credentials_configured):
+    """Reformats all of user's credentialed accounts to populate google_credentials_widget
+    dropdown's options. 
+
+    Args:
+        credentialed_accounts (str): user credentialed account to return credentials for
+        default_credentials_configured (boolean): This is True if google application-default
+        credentials are configured.
+
+    Returns:
+        dict: each key is a str of the users credentialed accounts and it maps to the 
+        same str credentialed account
+    """
+    accounts = credentialed_accounts
+    
+    accounts_dict = {}
+    for account in accounts:
+        accounts_dict[account] = account
+    if default_credentials_configured:
+        accounts_dict['default-credentials'] = 'default-credentials'
+    return accounts_dict
+
+
+def list_credentialed_accounts():
+    """Load all of user's credentialed accounts with ``gcloud auth list`` command.
+
+    Returns:
+        Sequence[str]: each key is a str of one of the users credentialed accounts
+
+    Raises:
+        sparkmagic.livyclientlib.BadUserConfigurationException: if account is not set or user
+        needs to run gcloud auth login or if gcloud is not installed.
+    """
+    accounts_json = ""
+    if os.name == "nt":
+        command = _CLOUD_SDK_WINDOWS_COMMAND
+    else:
+        command = _CLOUD_SDK_POSIX_COMMAND
+    try:
+        command = (command,) + _CLOUD_SDK_USER_CREDENTIALED_ACCOUNTS_COMMAND
+        accounts_json = subprocess.check_output(command, stderr=subprocess.STDOUT)
+        all_accounts = load_json_input(accounts_json)
+        credentialed_accounts = set()
+        active_account = None
+        for account in all_accounts:
+            try:
+                _cloud_sdk.get_auth_access_token(account['account'])
+                if account['status'] == 'ACTIVE':
+                    active_account = account['account']
+                credentialed_accounts.add(account['account'])
+            except:
+                pass
+        return credentialed_accounts, active_account
+    except (OSError) as caught_exc:
+        new_exc = BadUserConfigurationException(
+            "Gcloud is not installed. Install the Google Cloud SDK.")
```

Comment:
You may want to capture the exception message from `caught_exc`. There may be other reasons why gcloud cannot be invoked, e.g. if it's not on $PATH, permission denied, etc.

If stderr already has the error message printed, then nvm.

Reply:
After looking more closely, we would never get an error because we failed to obtain access token because this is just gcloud auth list command. So I changed it to just raise the exception and that gcloud cannot be invoked